JEP.md

```diff
@@ -0,0 +1,293 @@
+* **Jupyter extension developers** will be emitting events from both the server-side and the browser-side code.
+* **Jupyter operators** handle the infrastructure deployments and are interested in collecting various bits of events.
+* **Analysts** consume the final stored data and visualize and interpret it.
+
+### User Stories
+
+1. As a Jupyter user, I can view and understand the events being collected, so that I have visibility into what information is being collected.
+2. As a Jupyter user, I need to opt-in to event collection, so that I can choose what data about my usage is gathered.
+3. As a Jupyter extension developer, I can publish custom events using a well defined interface, from the browser as well as the server.
+4. As a Jupyter operator, I can configure events sinks for my deployment, so that I have control over where events are sent to.
+5. As a Jupyter operator, I can write a custom event sink for my deployment, so that I can control where event data is stored.
+6. As a Jupyter operator, I can decide what kinds and levels of events are going to be collected and stored.
+7. As a Jupyter operator, I can configure opt-in or opt-out for the users in my deployment.
+8. As an analyst,  I can clearly tell what the fields in the data represent so that I can interpret and make decisions from that data.
+
+## Proposed Enhancement
+
+The enhancement spans two categories - general concepts applicable across all Jupyter components, and the implementation for each concept for each Jupyter component.
+
+**General concepts:**
+
+* Event schema: The schema for a given event which is respected by all the Jupyter components across the event's lifecycle.
+* Event sinks: The "backends" which handle publishing the event to its final storage location.
+
+**Jupyter components:**
+
+* Jupyter server
+* JupyterLab
+* JupyterHub
+* Jupyter Classic
+
+## Detailed Explanation
+
+### Event Schema
+
+Each event is published conforming a given event schema and this is the contract followed by each step in the event lifecycle. These are referred to by all publishers regardless of whether the event comes from the server or the browser.
+
+Example event schema
+
+```json
+{
+    "name": "org.jupyter.kernel_lifecycle_event",
+    "version": 1,
+    "title": "Event emitted when a kernel lifecycle state changes",
+    "description": "Records each invocation of any command in JupyterLab",
+    "type": "object",
+    "args": {
+        "kernel_name": {
+            "type": "string",
+            "description": "Name of the kernel"
+        },
+        "state": {
+            "type": "string",
+            "description": "The new state of the kernel"
+        }
+    }
+}
+```
+
+Schema validation is done in the core telemetry framework that routes events from publishers to sinks, so that each event sink does not need to implement schema validation.
+
+**Open Questions**:
+
+* How do organizations deploy custom schemas?
+* Where are the schemas for public events stored?
```

Comment:
I chatted a little with @jasongrout in person last week about this. I think we should require schema URLs to be full absolute URLs that are resolvable. This is the only way the events can be 'self contained' - analysts can tell what the schema means, and operators can write services that process them in whatever way they want. This will also help answer the quesiton of 'where do public schemas get stored' and 'how do you have a private schema'.

Reply:
Would validating the schema at run-time require network access then? That would be a non-starter for many enterprise deployments. 

I'm open to suggestions for this section, especially.

Reply:
@jaipreet-s oooh, good point. That is definitely a non-starter. In JupyterHub / BinderHub now the schemas live in the package...

I guess there are two uses for schema:

1. Runtime validation at emit time
2. To figure out where things come from at analysis time

These probably need different ways to register and represent...

Reply:
> Runtime validation at emit time

The "package" approach makes most sense here, for performance reasons.

> To figure out where things come from at analysis time

Hmm, for v1 I'm *guessing* we could get by with just a package. WDYT?